Automattic/newspack-plugin
# Add Google OAuth2 authentication flow

### All Submissions:






### Changes proposed in this Pull Request:


 
Adds an OAuth2 authentication flow, as a preparation for the upcoming GAM integration.
Currently, Site Kit's authorisation is reused to access user's GA data. Site Kit does not support the scopes necessary to access GAM data, so Newspack has to roll its own authorisation for this purpose.  

UI: I have made it very lightweight purposefully, since this PR is quite large as it is. The UI should probably live in some kind of settings view. cc @thomasguillot 

### How to test the changes in this Pull Request:

1. Nothing should change unless the necessary environment variables are present, so have a look at the Dashboard, Campaigns Analytics, and Analytics wizard - everything should work as before
2. Create a [Google OAuth2 app](https://support.google.com/cloud/answer/6158849?hl=en). Set `<test-site->/wp-admin/admin.php?page=newspack` (e.g. `https://newspack.test/wp-admin/admin.php?page=newspack`) as an Authorized redirect URI.
3. Set client ID and secret of the app as `NEWSPACK_GOOGLE_OAUTH_CLIENT_ID` and `NEWSPACK_GOOGLE_OAUTH_CLIENT_SECRET` environment variables in `wp-config.php`
4. Visit the Dashboard, observe a new section above the wizard links - with an "Authorise" button
5. Click the button and complete the authorisation flow - observe a success message and information "Authorized Google as <email>" in the end
6. This was the happy path, now let's test how it can break. Remove the `_newspack_google_oauth` user meta from the DB
7. Reload the Dasboard, observe initial (pre-authorisation) state. Complete the authorisation from again.
8. Visit your [Google account's permissions screen](https://myaccount.google.com/permissions) and revoke permissions for the app
9. Again, observe the pre-authorisation state and complete the flow again.

### Other information:
